### DISCUSSION TITLE
Load drawio with files of different formats

### DISCUSSION THREAD

**Discussion by AngusParsonson** [marcveens/react-drawio#12] [Q&A] [answered]
Hi there,

Is there a way to load DrawIoEmbed with different file types, not just XML?

I want to load PNG, Visio files, etc., when the component is mounted. Currently, I can only load XML files by using the xml prop.

If I spin up DrawIoEmbed, I can drop these files into the editor, and it works nicely, so there might be a way to do this programmatically.

Thanks

**Comment by marcveens** [comment]
Hi Angus,

According to the draw.io docs it should be possible, I just have to find some time to check whether it's easily implemented within this plugin. I'll get back to you later!

**Comment by marcveens** [comment, marked as answer]
I just added an example of how to load a .png file, and transform it to the correct format:

https://github.com/marcveens/react-drawio/blob/f154102ec850d1ad1ea7a07479ebf7a483fcda7f/stories/DiagramsEmbed.stories.tsx#L53C1-L66C3

In case of .vsdx files, I don't know how draw.io handles them. There's not much to find online on how to parse those files. In the end, it should be an xml/base64 format one way or the other. I can't seem to parse that file type correctly, so unfortunately I'm not able to help with that.

**Comment by AngusParsonson** [comment]
Hey @marcveens, thanks for your response.

So is that just loading in a base64 string in the XML? 

Whenever I try to do this I don't see anything appear in the editor, or I get an error e.g. `Not a diagram file (error on line 1 at column 1: Start tag expected, '<' not found)`

Thanks,
Angus

**Reply by marcveens** [reply]
I had another look and managed to get it working!

See https://github.com/marcveens/react-drawio/blob/00fead6827dcc543dac74d7a7b069f5f35f05daf/stories/DiagramsEmbed.stories.tsx#L68C1-L83C3 for the Storybook implementation, and https://github.com/marcveens/react-drawio/blob/main/stories/hooks/useRemoteFile.tsx for the hook (or something similar) that you would need to include in your own project. 

Apparently it was in the embed documentation how to load a visio file; it needs to start with `data:application/vnd.visio;base64`. The hook makes sure the base64 is formatted correctly. 

Does that help you?

**Reply by marcveens** [reply]
I deployed the Storybook instance as a demo environment. See the working .vsdx example here: https://marcveens.github.io/react-drawio/?path=/story/components-drawioembed--with-remote-vsdx

**Reply by AngusParsonson** [reply]
Awesome thanks for your help @marcveens that works well!

I still can't seem to load png base64s but great to have the visio working.

**Reply by AngusParsonson** [reply]
Separate question @marcveens, do you know if it is possible to call the ref.export() without triggering any other event (save, load, etc.)? 

I was thinking I could use the .export method to do some conversion from .drawio format back to png.

**Reply by marcveens** [reply]
Unfortunately export will also trigger a save event, but I made an addition in version 0.1.6 (released just now).

As you can see in the console at https://marcveens.github.io/react-drawio/?path=/story/components-drawioembed--export-data there's now a `{ parentEvent: "export" }` property for the `onSave` and `onClose` events so you know if they are triggered by an export. This way you have a chance to ignore the save action when the export action is triggered.

**Reply by AngusParsonson** [reply]
Hey @marcveens, that's awesome thank you, although it looks to me like the parent event is still "export" when I click the blue save button in the top right. Same thing if I do file->save on the story example there, the parent event still seems to be "export".

**Reply by marcveens** [reply]
You're right... I was a little too fast with that release. I'm having a look if I can get it fixed now

**Reply by marcveens** [reply]
Alright, version 0.1.7 is a fact! The `parentEvent` is `export` when you call the action manually, and `save` when pressing the "save" button in the UI

**Reply by AngusParsonson** [reply]
Works a charm, thanks @marcveens!

**Reply by marcveens** [reply]
No problem! If you have some time, could you mark https://github.com/marcveens/react-drawio/discussions/12#discussioncomment-8732694 as an answer? That way other people can easily find it when they have a similar question. Thanks!

**Comment by AngusParsonson** [comment]
Is there a way to access the underlying `postMessage` function of draw.io? Perhaps I could call import with a visio file using postMessage?
